Unify directive dict structure

tgamblin opened the pull request:
Metadata created by directives on Spack packages isn't structured consistently, and it makes implementing similar functionality for different directives difficult and repetitive, e.g. with the `drop` directive in #35045.  It can also cause us to lose information about some parts of packages, e.g. how we lose the defaults of conditional variants in #38302.

Fundamentally, everything in a Spack package is *conditional*, which distinguishes Spack from other package managers.  The right way to key directive metadata is *by condition*, which prevents issues like the ones described above.  Also, storing by condition *seems* to make package loading faster, because there tend to be fewer top-level condition keys (from things like `with when(<condition>):` in packages than there are duplicated conditions in sub-dictionaries when you put conditions under some other top-level key (like the dependency name).

This PR converts Spack's metadata format as follows:

| Directive                   | Old                                                                 | New                                                                      |
|-----------------------------|---------------------------------------------------------------------|--------------------------------------------------------------------------|
| `conflicts`                 | <pre>conflict_spec:<br/>   [(when_spec, msg), ...]</pre>              | <pre>when_spec:<br/>   [ (conflict, msg), ... ]</pre>                      |
| `requires`                  | <pre>(requirement_spec, ...):<br/>   [(when_spec, policy, msg)]</pre> | <pre>when_spec:<br/>   [<br/>    ((requirement_spec, ...), policy, msg),<br/>    ...<br/>  ]</pre> |
| `depends_on`                | <pre>name:<br/>   { when_spec: Dependency }</pre>                     | <pre>when_spec:<br/>   { name: Dependency }</pre>                          |
| `provides`                  | <pre>provided_spec:<br/>   { when_spec, ... }</pre>                   | <pre>when_spec:<br/>   { provided_spec, ... }</pre>                        |
| `extends`                   | <pre>name:<br/>   (extendee_spec, unused_kwargs)</pre>                | <pre>when_spec:<br/>   [extendee_spec, ...]</pre>                          |
| `resource`<br/> (no change) | <pre>when_spec:<br/>   [Resource, ...]</pre>                          | <pre>when_spec:<br/>   [Resource, ...]</pre>                               |
| `patch`<br/> (no change)    | <pre>when_spec:<br/>   [Patch, ...]</pre>                            | <pre>when_spec:<br/>   [Patch, ...]</pre>                                |

Notably missing from the list above are variants and versions. Variants require quite a bit more work, and I am leaving them for a follow-on PR, as they require a lot more refactoring of packages and I think that will merit its own discussion.

Versions should likely be done as well, but we don't yet support `when` on them. Likely that should be done too, as we have users who would like to put conditions (particularly platform/OS conditions) on them. That is also a significant refactor for a future PR.

This PR does a few refactorings along with the changes above:

1.  `extendee_args`: this method is unused, and we no longer pass `kwargs` to `extends()` directive.
2. Code and concretizer refactors for all the code in Spack that *uses* the directives above. 
4. Add typing information to `PackageBase` for all the dictionaries above, ensuring that we parse all data that needs to be a `Spec` early (thus failing early), and that we get consistent metadata dictionaries on our package classes.
5.  Make `_make_when_spec()` internal to `directives.py` and don't use it in the solver.  We don't need to make specs in the solver if we guarantee (via `mypy` that they're constructed when packages are parsed.

tgamblin committed:
refactor: Index dependency metadata by `when` spec

Part 1 of making all package metadata indexed by `when` condition. This
will allow us to handle all the dictionaries on `PackageBase` consistently.

Convert the current dependency dictionary structure from this:

    { name: { when_spec: [Dependency ...] } }

to this:

    { when_spec: { name: [Dependency ...] } }

On an M1 mac, this actually shaves 5% off the time it takes to load all
packages, I think because we're able to trade off lookups by spec key
for more lookups by name.

tgamblin committed:
refactor: Index conflicts by `when` spec

Part 2 of reworking all package metadata to key by `when` conditions.

Changes conflict dictionary structure from this:

    { conflict_spec: [(when_spec, msg), ...] }

to this:

    { when_spec: [(conflict_spec, msg), ...] }

Also attempts to consistently name the variables used to iterate over conflict
dictionaries.

tgamblin committed:
refactor: Index requirements by `when` spec

Part 3 of reworking all package metadata to key by `when` conditions.

Changes conflict dictionary structure from this:

    { (requirement_spec, ...): [(when_spec, policy, msg)] }

to this:

    { when_spec: [((requirement_spec, ...), policy, msg), ...] }

tgamblin committed:
refactor: Index provided virtuals by `when` spec

Part 4 of reworking all package metadata to key by `when` conditions.

Changes conflict dictionary structure from this:

    { provided_spec: {when_spec, ...} }

to this:

    { when_spec: {provided_spec, ...} }

adamjstewart commented:
Why don't versions list URLs anymore? Also, the "preferred version" of `py-cython` is broken, both on this PR and on develop.

tgamblin commented:
> Why don't versions list URLs anymore?

Confused -- are you just seeing that I only screenshotted part of the output?  I didn't screenshot the version part.

bonachea commented:
@tgamblin this looks great! Thanks for finally addressing #23831

adamjstewart commented:
> > Why don't versions list URLs anymore?
> 
> Confused -- are you just seeing that I only screenshotted part of the output? I didn't screenshot the version part.

I checked out the branch locally and `spack info ...` shows None for all URLs in all packages.

tgamblin commented:
@adamjstewart will fix -- that's a bug (still in draft 😅)

tgamblin committed:
refactor: make `_make_when_spec()` private to `directives.py`

`make_when_spec()` was being used in the solver, but it has semantics that are specific
to parsing when specs from `package.py`. In particular, it returns `None` when the
`when` spec is `False`, and directives are responsible for ignoring that case and not
adding requirements, deps, etc. when there's an actual `False` passed in from
`package.py`.

In `asp.py`, we know that there won't ever be a raw boolean when spec or constraint, so
we know we can parse them without any of the special boolean handling. However, we
should report where in the file the error happened on error, so this adds some parsing
logic to extract the `mark` from YAML and alert the user where the bad parse is.

- [x] refactor `config.py` so that basic `spack_yaml` mark info is in its own method
- [x] refactor `asp.py` so that it uses the smarter YAML parsing routine
- [x] refactor `asp.py` so that YAML input validation for requirements is done up front

alalazo commented:
I started reviewing the PR by running some benchmark. The `setup` phase is _slightly slower_ than on `develop`:

develop: 5d48130cc693301553771f29ea7463eebe3219d6
PR: 1effa2958dd70357f098d993efa5fc114c071cc1

[radiuss.runtime.csv](https://github.com/spack/spack/files/13817856/radiuss.runtime.csv)
[radiuss.develop.csv](https://github.com/spack/spack/files/13817857/radiuss.develop.csv)
[radiuss.txt](https://github.com/spack/spack/files/13817858/radiuss.txt)

![radiuss](https://github.com/spack/spack/assets/4199709/1a2b964e-8662-4592-a2a6-a210fb77e019)

alalazo commented on lib/spack/spack/directives.py:
```diff
@@ -455,8 +463,14 @@ def _execute_version(pkg, ver, **kwargs):
 
-def _depends_on(pkg, spec, when=None, type=dt.DEFAULT_TYPES, patches=None):
-    when_spec = make_when_spec(when)
+def _depends_on(
+    pkg: "spack.package_base.PackageBase",
+    spec: SpecType,
+    when: WhenType = None,
+    type: DepType = dt.DEFAULT_TYPES,
+    patches: PatchesType = None,
+):
```
(minor) Asking for explicit keyword arguments usually makes the calling site more legible:
```suggestion
def _depends_on(
    pkg: "spack.package_base.PackageBase",
    *,
    spec: SpecType,
    when: WhenType = None,
    type: DepType = dt.DEFAULT_TYPES,
    patches: PatchesType = None,
):
```

tgamblin replied:
I agree -- however I think `spec` should be positional. Did you mean to include it?

alalazo replied:
Yep, apologies, it's a typo

tgamblin replied:
fixed

alalazo commented on lib/spack/spack/directives.py:
```diff
@@ -509,7 +524,7 @@ def _depends_on(pkg, spec, when=None, type=dt.DEFAULT_TYPES, patches=None):
 
 @directive("conflicts")
-def conflicts(conflict_spec, when=None, msg=None):
+def conflicts(conflict_spec: SpecType, when: WhenType = None, msg: Optional[str] = None):
```
I would also like to use explicit kwargs on directives, but that requires a deprecation + warning step first (so maybe better leave it for a following PR).

tgamblin replied:
It looks like we have a very small number of packages that don't use `when=` with `conflicts()`, and none that omit `msg=`.  But yeah let's leave this for a follow-on.

alalazo commented on lib/spack/spack/directives.py:
```diff
@@ -561,14 +581,14 @@ def depends_on(spec, when=None, type=dt.DEFAULT_TYPES, patches=None):
 
 @directive(("extendees", "dependencies"))
-def extends(spec, type=("build", "run"), **kwargs):
+def extends(spec, when=None, type=("build", "run"), patches=None):
```
(minor) Technically, if we want to be non-breaking, we need to swap the arguments:
```suggestion
def extends(spec, type=("build", "run"), when=None, patches=None):
```
However, I guess the number of users having a call to `extends` with only positional arguments is very low...

tgamblin replied:
I tend to agree. I don't think this is a major change.

alalazo commented on lib/spack/spack/directives.py:
```diff
@@ -621,36 +641,40 @@ def _execute_provides(pkg):
-            if provided_spec not in pkg.provided:
-                pkg.provided[provided_spec] = set()
-            pkg.provided[provided_spec].add(when_spec)
+            provided_set = pkg.provided.setdefault(when_spec, set())
+            provided_set.add(provided_spec)
```
Why using a `set` here, instead of a list of virtual specs (eventually constrained if multiple directives refer to the same virtual under the same condition)?

alalazo replied:
Is there any technical reason for this :point_up:

tgamblin replied:
I guess my question would be why not a `set`?  The virtuals provided need to be unique... why use a list and list the same thing twice?

alalazo replied:
> why use a list and list the same thing twice?

I'm not asking that, see "(eventually constrained if multiple directives refer to the same virtual under the same condition)".

What I'm wondering is why storing:
```python
provides("mpi", when="+foo")
provides("mpi@2:", when="+foo")
provides("blas", when="+foo")
```
as `{"mpi", "mpi@2:", "blas"}` rather than just `["mpi@2:", "blas"]`.

tgamblin replied:
This is similar to what Harmen has been asking. `provides` is not the same as `depends_on` — with dependencies, you need to intersect all the constraints on each dep. With `provides()` you need to union them.

In the example above, you’re saying the thing provides `mpi` whatever version when `+foo`, and `mpi` 2 or higher as well. The union of those is `mpi`, so we *could* add a union operation and use it here; we just don’t happen to have one.

most packages don’t write their provides this way — you typically want something like what we use for MPI:

```
provides(“mpi@:1”, when=“@1:”)
provides(“mpi@:2”, when=“@3:”)
```

This uses upper bounds. version 2 introduces mpi 1 support and version 3 introduces mpi 2 support (but keeps mpi 1 support as well — the provides are unioned where the whens overlap).

Lower (right-open) bounds don’t really make sense for provides() — that would mean you provide every future version of something… which you can’t possibly do because future versions aren’t released.

Providers are producers (providing all features/versions declared) not consumers (for whom you must satisfy all requirements declared); they’re on the other side of the depends_on directive here so it has to be flipped.

alalazo replied:
> Lower (right-open) bounds don’t really make sense for provides() — that would mean you provide every future version of something… which you can’t possibly do because future versions aren’t released.

This is slightly OT, but I agree with that. This is also inline with what I was proposing yesterday at the meeting: enforce virtual packages to be defined and have a list of available versions. Once we have that, we can also have a proper audit message for these directives, if they ever omit an upper bound.

alalazo commented on lib/spack/spack/directives.py:
```diff
@@ -621,36 +641,40 @@ def _execute_provides(pkg):
-            (only required for URL patches)
-        archive_sha256 (str): sha256 sum of the *archive*, if the patch
-            is compressed (only required for compressed URL patches)
+        sha256: sha256 sum of the patch, used to verify the patch (only required for URL patches)
```
We can remove the `Keyword Args:` above

tgamblin replied:
done

alalazo commented on lib/spack/spack/audit.py:
```diff
@@ -694,25 +694,21 @@ def _unknown_variants_in_directives(pkgs, error_cls):
-                        pkg_cls, vrn, directive="patch", error_cls=error_cls
-                    )
+        # Check "provides" directive
+        for when_spec in pkg_cls.provided.keys():
```
Should be the default:
```suggestion
        for when_spec in pkg_cls.provided:
```

tgamblin replied:
done

alalazo commented on lib/spack/spack/audit.py:
```diff
@@ -736,70 +732,60 @@ def _issues_in_depends_on_directive(pkgs, error_cls):
                     ]
                     errors.append(error_cls(summary=summary, details=details))
 
-                for s in (dependency_spec, when):
-                    if s.virtual and s.variants:
-                        summary = f"{pkg_name}: virtual dependency cannot have variants"
-                        details = [
-                            f"remove variants from '{str(s)}' in depends_on directive",
-                            f"in {filename}",
-                        ]
-                        errors.append(error_cls(summary=summary, details=details))
```
(Reverted audit) This check was removed, and should stay

tgamblin replied:
restored

alalazo commented on lib/spack/spack/package_base.py:
```diff
@@ -432,6 +432,43 @@ def remove_files_from_view(self, view, merge_map):
+NameWhenDict = Dict[str, Dict[spack.spec.Spec, List[Any]]]
+
+
+def _by_name(
```
If I get this correctly, the function converts to the old `dependencies` format. Where is this needed? Can we avoid it?

tgamblin replied:
there are some loops that are easier to do by name, and some that are easier to do by `when`.  It'll be more obvious when we do variants but I think it's needed (and useful) to have both.

alalazo commented on lib/spack/spack/solver/asp.py:
```diff
@@ -1130,6 +1135,19 @@ def __iter__(self):
         return iter(self.data)
 
 
+def parse_spec_from_yaml_string(string):
+    """Parse a spec from YAML and add file/line info to errors, if it's available."""
+    try:
+        spec = spack.spec.Spec(string)
+        return spec
+    except spack.parser.SpecSyntaxError as e:
+        mark = spack.config.get_mark_from_yaml_data(string)
+        if mark:
+            msg = f"{mark.name}:{mark.line + 1}: {str(e)}"
+            raise spack.parser.SpecSyntaxError(msg) from e
+        raise e
```
This should be moved to a module dealing with YAML or config, seems out of place in the `spack.solver` package.

tgamblin replied:
moved to config

alalazo commented on lib/spack/spack/test/concretize_requirements.py:
```diff
@@ -856,6 +856,8 @@ def test_skip_requirement_when_default_requirement_condition_cannot_be_met(
 
 def test_requires_directive(concretize_scope, mock_packages):
     compilers_yaml = pathlib.Path(concretize_scope) / "compilers.yaml"
+
+    # NOTE: target is omitted here so that the test works on aarch64, as well.
```
This can probably be extracted as a smaller fix

tgamblin replied:
done in #41943

alalazo reviewed (changes requested):
I gave a first quick pass. I think an audit (variants on virtuals) has been accidentally deleted in this PR.

alalazo commented:
Q: can we extract the commit on `extendee_args` as a separate PR ?

tgamblin commented:
@alalazo 
> I started reviewing the PR by running some benchmark. The setup phase is slightly slower than on develop

I think the slowdown is likely that after adding type information, I started parsing specs earlier (during import). I think this is actually a good thing, as it will detect errors in packages sooner.  The same changes are standalone in #41630... maybe we should compare to that?

tgamblin commented:
> Q: can we extract the commit on extendee_args as a separate PR ?

done in #41942.

tgamblin commented:
@alalazo ok I think the review comments are addressed

alalazo reviewed (changes requested):
LGTM. We need to rebase this and merge it. I have a question, mainly for my knowledge.

tgamblin commented:
@alalazo rebased -- see comment in thread above

alalazo reviewed (approved):
There is a minor thread on-going, but it's about an internal implementation detail that, in case, can be changed in a later PR.
